### PR TITLE
An optimisation of underapproximation

### PR DESCRIPTION
**Pull request by vhavlena**
A minor change of co-finite check to allow a wider range of constraint to pass the underapproximation.

### CODE REVIEW

**Comment by vhavlena** [comment]
Results on all benchmarks.
```
# of formulae: 82972
###################################################################################
####                                   Table 1                                 ####
###################################################################################
| method                     |    max |     mean |   median |   std. dev |   TO+MO+ERR |   unknowns |
|----------------------------|--------|----------|----------|------------|-------------|------------|
| z3-noodler-482f3c0-387babd | 109.13 | 0.500252 |     0.02 |    3.69626 |        2028 |       2012 |
| cvc5-1.0.8                 | 119.36 | 0.515113 |     0.02 |    5.2351  |        1723 |          2 |
| z3-4.12.2                  | 119.29 | 1.49018  |     0.04 |    8.17615 |        2623 |        123 |
| ostrich-70d01e2d2-parikh   | 577.7  | 5.35069  |     2.47 |   11.7293  |       25162 |          0 |
| z3-noodler-feada45-a57f582 | 116.1  | 0.494545 |     0.02 |    3.67665 |        2198 |       2011 |
```

On `kaluza` 160 TOs less than devel.

kaluza

**Review by jurajsic — APPROVED** [review]
Just some small stuff

**Review by Adda0 — APPROVED** [review]
Other than the already mentioned, seems fine to me.

File: src/smt/theory_str_noodler/aut_assignment.h [hunk 1]
```diff
@@ -252,6 +247,15 @@ namespace smt::noodler {
          */
         LenNode get_lengths(const BasicTerm& var) const;
 
+        /**
+         * @brief Get the lengths formula representing all possible lengths of the automaton for @p var and corresponding NFA @p aut.
+         * 
+         * @param aut Automaton
+         * @param var Variable
+         * @return LenNode
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/aut_assignment.h]: ok, fixed

File: src/smt/theory_str_noodler/formula_preprocess.cpp [hunk 2]
```diff
@@ -1381,12 +1381,15 @@ namespace smt::noodler {
...
-                    LenNode left = LenNode(var);
-                    LenNode eq = LenNode(LenFormulaType::EQ, {left, right});
-                    this->add_to_len_formula(LenNode(LenFormulaType::NOT, {eq}));
+                if(this->aut_ass.is_co_finite(var)) {
+                    auto alphabet =  this->aut_ass.get_alphabet(false);
+                    mata::OnTheFlyAlphabet mata_alphabet{};
+                    for (const auto& symbol : alphabet) {
+                        mata_alphabet.add_new_symbol(std::to_string(symbol), symbol);
+                    }
+                    mata::nfa::Nfa aut_compl = mata::nfa::complement(*(this->aut_ass.at(var)), mata_alphabet);
```
[the remark this answers is not in the document]
Comment (vhavlena) [reply src/smt/theory_str_noodler/formula_preprocess.cpp]: 1) good point. Should be fixed now
2) true, but I don't want to make a side effect in `is_co_finite`.

**Pull request merged**